include/glow/Graph/Node.h

```diff
@@ -352,6 +365,84 @@ class NodeWalker {
   virtual ~NodeWalker() = default;
 };
 
+/// Helper class to walk through the specific uses of a NodeValue.
+/// This class is built on top of the regular users-list (Node::getUsers)
+/// but filters out the uses that don't affect the desired NodeValue.
+template <bool is_const_iter = false>
+class NodeValueIteratorImpl
+    : public std::iterator<std::forward_iterator_tag, NodeUse> {
+public:
+  /// Base type of the iterator.
+  using iterator =
+      typename std::conditional<is_const_iter,
+                                std::list<NodeUse>::const_iterator,
+                                std::list<NodeUse>::iterator>::type;
+  /// Type of the NodeValue that this iterator is filtering for.
+  using NodeValueTy = typename std::conditional<is_const_iter, const NodeValue,
+                                                NodeValue>::type;
+  /// Type of the NodeUse that this iterator should return when dereferenced.
+  using NodeUseTy =
+      typename std::conditional<is_const_iter, const NodeUse, NodeUse>::type;
+
+private:
+  /// NodeValue that this iterator tracks.
+  NodeValueTy &parent_;
+  /// Actual iterator on the users-list.
+  /// \invariant if it_ points to a valid iterator, then the NodeValue it
+  /// references (via the NodeUse) is equal to parent_.
+  iterator it_;
+
+  /// Convenient method to get the end iterator of the users list that this
+  /// iterator walks.
+  iterator getEnd() const { return parent_.getNode()->getUsers().end(); }
+
+  /// Check if \p it_ points to a NodeUse that references \p parents_.
+  bool hasSameParent() const {
+    assert(it_ != getEnd() && "Cannot check invalid iterator");
+    // A users-list should be for one node.
+    // If this assert breaks, that means the input list is broken,
+    // or this iterator is not used as it was intended: to walk
+    // through a users-list.
+    assert(it_->get()->getNode() == parent_.getNode() &&
+           "Iterator points to a list with different parent?!");
+    return it_->get()->getResNo() == parent_.getResNo();
+  }
+
+public:
+  NodeValueIteratorImpl(NodeValueTy &parent, iterator it)
+      : parent_(parent), it_(it) {
+    if (it_ != getEnd() && !hasSameParent()) {
+      ++(*this);
+    }
+    assert((it_ == getEnd() || hasSameParent()) &&
+           "operator++ should return the next valid iterator");
+  }
+
+  const NodeValueTy &getParent() const { return parent_; }
```

Comment:
This doesn't seem to be used -- should it be removed, or be made private? Otherwise I think it should be included in the unit test.

Reply:
Good catch, I meant to test it :)

Reply:
Finally I decided to drop it. I meant to add some asserts in the test, but I actually already did the checks I wanted within the class (::hasSameParent).